amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,549 @@
+        if self.toolchain == "Apicula":
+            return self._apicula_command_templates
+        elif self.toolchain == "Gowin":
+            return self._gowin_command_templates
+        assert False
+
+    def add_clock_constraint(self, clock, frequency):
+        super().add_clock_constraint(clock, frequency)
+        clock.attrs["keep"] = "true"
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "sys_clk0":
+            return Clock(self.osc_freq / self.osc_div)
+        # Use the defined Clock resource.
+        return super().default_clk_constraint
+
+    def create_missing_domain(self, name):
+        if name == "sync" and self.default_clk is not None:
+            m = Module()
+            if self.default_clk == "sys_clk0":
+                props = self._osc_props
+                osc = self._oscillator
+                if (not isinstance(self.osc_div, int)
+                    or self.osc_div < props["div_min"]
+                    or self.osc_div > props["div_max"]
+                    or (self.osc_div % props["div_step"]) != 0):
+                    raise ValueError("OSC divider (osc_div) must be an integer between {} and {} in steps of {}, not {!r}".format(
+                        props["div_min"], props["div_max"], props["div_step"], self.osc_div))
+
+                clk_i = Signal()
+                if osc is "OSCZ":
+                    m.submodules += Instance(osc,
+                                             p_FREQ_DIV=self.osc_div,
+                                             i_OSCEN=Const(1),
+                                             o_OSCOUT=clk_i)
+                elif osc is "OSCO":
+                    # TODO: Make use of regulator configurable
+                    m.submodules += Instance(osc,
+                                             p_REGULATOR_EN=Const(1),
+                                             p_FREQ_DIV=self.osc_div,
+                                             i_OSCEN=Const(1),
+                                             o_OSCOUT=clk_i)
+                elif osc is "OSCF":
+                    m.submodules += Instance(osc,
+                                             p_FREQ_DIV=self.osc_div,
+                                             o_OSCOUT30M=None,
+                                             o_OSCOUT=clk_i)
+                else:
+                    m.submodules += Instance(osc,
+                                             p_FREQ_DIV=self.osc_div,
+                                             o_OSCOUT=clk_i)
+
+            else:
+                clk_i = self.request(self.default_clk).i
+
+            if self.default_rst is not None:
+                rst_i = self.request(self.default_rst).i
+            else:
+                rst_i = Const(0)
+
+            m.domains += ClockDomain("sync")
+            m.d.comb += ClockSignal("sync").eq(clk_i)
+            m.submodules.reset_sync = ResetSynchronizer(rst_i, domain="sync")
```

Comment:
After reviewing the GW1N datasheet I think this code should be done like Lattice ECP5, using GSR/SGSR.

Reply:
No idea what that is about, but can find out. Need a break though.

Reply:
Thanks! I'm very happy with the progress we're making.

Reply:
There is only asynchronous `GSR` on all Gowin chips. No `SGSR`. I'll use something along the lines of what is done for EPC5. The ECP5 `FD1S3AX` should be equivalent to Gowin `DFF`, right?

Reply:
Oh, if we don't have `SGSR` then I think the current implementation is fine. It would only be because of `SGSR`.

Reply:
Thinking about it, I wonder if the `GSR` implemenation actually works with nextpnr-gowin. I remember there was some additional work to make it happen for ECP5...

Reply:
There were no complaints at compile time, but I did not test with hardware. Now back to using previous implementation.